nixos/modules/services/system/swapspace.nix

```diff
@@ -0,0 +1,103 @@
+    package = mkPackageOption pkgs "swapspace" { };
+    settings = mkOption {
+      type = types.submodule {
+        options = {
+          swappath = mkOption {
+            type = types.str;
+            default = "/var/lib/swapspace";
+            description = "Location where swapspace may create and delete swapfiles";
+          };
+          lower_freelimit = mkOption {
+            type = types.ints.between 0 99;
+            default = 20;
+            description = "Lower free-space threshold: if the percentage of free space drops below this number, additional swapspace is allocated";
+          };
+          upper_freelimit = mkOption {
+            type = types.ints.between 0 100;
+            default = 60;
+            description = "Upper free-space threshold: if the percentage of free space exceeds this number, swapspace will attempt to free up swapspace";
+          };
+          freetarget = mkOption {
+            type = types.ints.between 2 99;
+            default = 30;
+            description = ''
+              Percentage of free space swapspace should aim for when adding swapspace.
+              This should fall somewhere between lower_freelimit and upper_freelimit.
+            '';
+          };
+          min_swapsize = mkOption {
+            type = types.str;
+            default = "4m";
+            description = "Smallest allowed size for individual swapfiles.";
+          };
+          max_swapsize = mkOption {
+            type = types.str;
+            default = "2t";
+            description = "Greatest allowed size for individual swapfiles.";
+          };
+          cooldown = mkOption {
+            type = types.ints.unsigned;
+            default = 600;
+            description = ''
+              Duration (roughly in seconds) of the moratorium on swap allocation that is instated if disk space runs out, or the cooldown time after a new swapfile is successfully allocated before swapspace will consider deallocating swap space again.
+              The default cooldown period is about 10 minutes.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Config file for swapspace.
+        See the options here: <https://github.com/Tookmund/Swapspace/blob/master/swapspace.conf>
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+    systemd.services.swapspace = {
+      after = [
+        "local-fs.target"
+        "swap.target"
+      ];
+      requires = [
+        "local-fs.target"
+        "swap.target"
+      ];
+      wantedBy = [ "multi-user.target" ];
+      documentation = [ "man:swapspace(8)" ];
+      description = "Swapspace, a dynamic swap space manager";
+      serviceConfig = {
+        ExecStart = "${lib.getExe cfg.package} -c ${configFile}";
+        StateDirectory = lib.mkIf (cfg.settings.swappath == "/var/lib/swapspace") "swapspace";
```

Comment:
@Luflosi does this make sense? If the swappath is set to something non-default, mkif fails and statedirectory will be null.

Reply:
using systemd.tmpfile.rules, since it could be outside `/var/lib/`

Reply:
```shell
$ sudo ls -l /var/lib | rg swapspace
drwx------  2 root            root            4096 Oct 15 12:04 swapspace
$ ls -l /var/lib/swapspace
"/var/lib/swapspace": Permission denied (os error 13)
```

Reply:
I can think of a use case, where your / is not the fastest ssd you have plugged in.

Reply:
It does work outside `/var/lib` I've tried `/shed/swamp` as the `settings.swappath` and it works as expected with correct permissions.

```shell
$ swapon --show
NAME           TYPE       SIZE  USED PRIO
/dev/nvme0n1p4 partition 16.2G 14.4G   -2
/dev/zram0     partition 13.8G 13.8G    5
/shed/swamp/1  file       6.7G    0B   -3
/shed/swamp/2  file       7.7G    0B   -4
```